component/kafka/simple/component.go

```diff
@@ -0,0 +1,359 @@
+	partitions, err := consumer.Partitions(c.topic)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get partitions: %w", err)
+	}
+
+	// When kafka cluster is not fully initialized, we may get 0 partitions.
+	if len(partitions) == 0 {
+		return nil, errors.New("got zero partitions")
+	}
+
+	offsets, err := c.getStartingOffsets(ctx, client, consumer, partitions)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get starting offsets: %w", err)
+	}
+
+	pcs := make(map[int32]sarama.PartitionConsumer, len(partitions))
+	for _, partition := range partitions {
+		pc, err := consumer.ConsumePartition(c.topic, partition, offsets[partition])
+		if nil != err {
+			return nil, fmt.Errorf("failed to get partition consumer: %w", err)
+		}
+		pcs[partition] = pc
+	}
+
+	if c.notificationOnceReachingLatestOffset() {
+		c.startingOffsets = offsets
+		err := c.setLatestOffsets(client, partitions)
+		if err != nil {
+			return nil, fmt.Errorf("failed to set latest offsets: %w", err)
+		}
+	}
+
+	return pcs, nil
+}
+
+func (c *Component) getStartingOffsets(ctx context.Context, client sarama.Client, consumer sarama.Consumer, partitions []int32) (map[int32]int64, error) {
+	if c.durationBasedConsumer {
+		return c.getTimeBasedOffsetsPerPartition(ctx, client, consumer, partitions)
+	}
+	return c.getOffsetsFromInitialOffset(client, partitions)
+}
+
+func (c *Component) getOffsetsFromInitialOffset(client sarama.Client, partitions []int32) (map[int32]int64, error) {
+	offsets := make(map[int32]int64)
+	for _, partitionID := range partitions {
+		offset, err := client.GetOffset(c.topic, partitionID, c.saramaConfig.Consumer.Offsets.Initial)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get offset for partition %d: %w", partitionID, err)
+		}
+		offsets[partitionID] = offset
+	}
+
+	return offsets, nil
+}
+
+func (c *Component) getTimeBasedOffsetsPerPartition(ctx context.Context, client sarama.Client, consumer sarama.Consumer, partitions []int32) (map[int32]int64, error) {
+	dkc, err := newDurationKafkaClient(client, consumer, c.saramaConfig.Net.DialTimeout)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Kafka duration client: %w", err)
+	}
+
+	dc := newDurationClient(dkc, partitions)
+	offsets, err := dc.getTimeBasedOffsetsPerPartition(ctx, c.topic, time.Now().Add(-c.durationOffset), c.timeExtractor)
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve duration offsets per partition: %w", err)
+	}
+
+	return offsets, nil
+}
+
+func (c *Component) newConsumer() (sarama.Client, sarama.Consumer, error) {
+	client, err := sarama.NewClient(c.brokers, c.saramaConfig)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create client: %w", err)
+	}
+
+	consumer, err := sarama.NewConsumerFromClient(client)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create simple consumer: %w", err)
+	}
```

Comment:
Since the client is created successfully at this point, we need to `closeClient(client)` in case of a consumer error.